blueprints/flexible-topologies.md

```diff
@@ -0,0 +1,395 @@
+                "parents": []
+            }
+        ]
+    }
+}
+```
+
+##### `DELETE /topologies?name=foo`
+
+response JSON:
+```JSON
+{
+    "alerts": [
+        {
+            "text": "topology was deleted successfully",
+            "level": "success"
+        }
+    ]
+}
+```
+
+##### `/deliveryservices` endpoints
+
+All relevant Delivery Service APIs will have their JSON request and response objects modified to include a new `topology` field which references the name of the topology it's assigned to:
+```JSON
+{
+    ...
+    "topology": "foo"
+}
+```
+
+##### The various `/snapshot` endpoints
+
+The various `/snapshot` endpoints will need to be updated to include new Topologies data along with their associations to Delivery Services in the `CRConfig.json` snapshot. The data should only include the `EDGE_LOC` cachegroups of the Topologies, because those are all Traffic Router needs.
+
+##### Various endpoints that are affected by cachegroup parentage or deliveryservice-server assignment
+
+API endpoints that do things such as the following may need to be updated to take Topology-based Delivery Service assignment and parentage into account:
+- assign a Delivery Service to a server (or vice versa)
+- return the servers that are assigned to a Delivery Service (or vice versa)
+- perform an operation on "child" cachegroups -- like queueing updates on "child" caches when changing the status of a "parent" cache
+
+#### Client Impact
+
+New Go client methods will be added for the `/topologies` endpoints in order to write TO API tests for the new endpoints. The `/deliveryservices` client methods won't need modified as the new `Topology` field will simply be added to the `DeliveryService` struct. New client methods for the Python client will also be added for each of the new `/topologies` endpoints.
+
+#### Data Model Impact
+
+##### Go structs
+
+New `Topology` and `TopologyNode` structs will be added for the `/topologies` endpoints, mapping directly to the JSON request bodies in the REST API Impact section. The `DeliveryService` struct will be updated with a new `Topology` field which is the name of the Topology the Delivery Service is assigned to.
+
+##### Traffic Ops Database
+
+A new `topology` table will be created:
+
+| column      | type | modifiers    |
+| ----------- | ---- | ------------ |
+| name        | text | not null, PK |
+| description | text |              |
+
+A new `topology_cachegroup` table will be created to model the association of cachegroups to topologies:
+
+| column     | type | modifiers                                       |
+| ---------- | ---- | ----------------------------------------------- |
+| id         | int  | not null, PK                                    |
+| topology   | text | not null, FK: references topology(name)         |
+| cachegroup | text | not null, FK: references cachegroup(short_name) |
+
+**Constraints**:
+- unique (topology, cachegroup) -- a cachegroup can only be in a Topology once.
```

Comment:
i thought they could be in a topology more than once...

Reply:
nevermind, the design supports it but for now a cachegroup will only appear in a topology once max. got it.